src/Lazy/transpose.ts

```diff
@@ -0,0 +1,132 @@
+import { isAsyncIterable, isIterable, toIterator } from "../_internal/utils";
+import isUndefined from "../isUndefined";
+import max from "../max";
+import nth from "../nth";
+import pipe from "../pipe";
+import size from "../size";
+import toArray from "../toArray";
+import type ReturnZipType from "../types/ReturnZipType";
+import type { UniversalIterable } from "../types/Utils";
+import filter from "./filter";
+import map from "./map";
+import toAsync from "./toAsync";
+
+function* sync(
+  iterable: Iterable<Iterable<unknown>>,
+): IterableIterator<Iterable<unknown>> {
+  const maxLen = pipe(iterable, map(size), max);
+
+  for (let index = 0; index < maxLen; index++) {
+    yield pipe(
+      iterable,
+      map(nth(index)),
+      filter((value) => !isUndefined(value)),
+      toArray,
+    );
+  }
+}
+
+function async(
+  iterable: Iterable<UniversalIterable>,
+): AsyncIterableIterator<UniversalIterable> {
+  const iterators = toArray(map(toIterator, iterable as any));
+
+  return {
+    [Symbol.asyncIterator]() {
+      return this;
+    },
+    async next(_concurrent) {
+      const headIterators = await pipe(
+        toAsync(iterators),
+        map((iterator) => iterator.next(_concurrent)),
+        toArray,
+      );
+
+      const done = headIterators.every(({ done }) => done);
+
+      if (done) return { done: true, value: undefined };
+
+      return {
+        done: false,
+        value: pipe(
+          headIterators,
+          map(({ value }) => value),
+          filter((value) => !isUndefined(value)),
+          toArray,
+        ),
+      };
+    },
+  };
+}
+
+/**
+ * The transpose function takes multiple iterators as input and returns a transposed 2D array—swapping rows with columns.
+ *
+ * Unlike the existing zip function, transpose behaves as shown in the example below:
+ *
+ * @example
+ * ```ts
+ * const iter = transpose([1, 2], [3], [], [4, 5, 6], [7], [8, 9]);
+ * iter.next() // {done:false, value: [1, 3, 4, 7, 8]}
+ * iter.next() // {done:false, value: [2, 5, 9]}
+ * iter.next() // {done:false, value: [6]}
+ * iter.next() // {done:true, value: undefined}
```

Comment:
The general usage of transpose is as follows,
So it would be good to include the example below.

```
R.transpose([[1, 'a'], [2, 'b'], [3, 'c']]) //=> [[1, 2, 3], ['a', 'b', 'c']]
R.transpose([[1, 2, 3], ['a', 'b', 'c']]) //=> [[1, 'a'], [2, 'b'], [3, 'c']]
```
https://ramdajs.com/docs/#transpose

I think it would be good to write the example you gave in the form of an exceptional case.